### PR TITLE
Generate representative applications for 2020 & 2021

### PR DESCRIPTION
**Pull request by tijmenb**
## Context

We're improving the test application generation.

## Changes proposed in this pull request

This refactors the test application code a bit to make sure we generate applications for 2020 in terminal states.

## Guidance to review

Check out the review app:

<>

Do we have all relevant application covered?

## Link to Trello card

https://trello.com/c/QiEcDyYo

## Things to check

- [X] This code does not rely on migrations in the same Pull Request
- [X] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [X] API release notes have been updated if necessary
- [X] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/workers/generate_test_applications.rb [hunk 1]
```diff
@@ -1,50 +1,63 @@
...
-  def create(states:, apply_again: false, course_full: false)
-    @test_applications.create_application(
-      courses_to_apply_to: @courses_to_apply_to,
+private
+
+  def create(recruitment_cycle_year: nil, states:, apply_again: false, course_full: false)
+    TestApplications.new.create_application(
       states: states,
+      recruitment_cycle_year: recruitment_cycle_year,
+      courses_to_apply_to: recruitment_cycle_year == 2020 ? @courses_to_apply_to_in_2020 : @courses_to_apply_to_in_2021,
```
Comment (mnacos) [inline app/workers/generate_test_applications.rb:46]:
```suggestion
      courses_to_apply_to: recruitment_cycle_year == 2020 ? courses_to_apply_to_in_2020 : courses_to_apply_to_in_2021,
```
To make sure these instance variables have been initialised.

File: lib/tasks/generate_test_data.rake [hunk 2]
```diff
@@ -1,19 +1,6 @@
 desc 'Generate test applications for existing providers'
 task generate_test_applications: :environment do
-  GenerateTestApplications.new(for_year: :previous_year).perform
-
-  GetApplicationChoicesReadyToRejectByDefault.call.find_each do |choice|
```
Comment (mnacos) [inline lib/tasks/generate_test_data.rake:5]: Removing this explicit rejection step between years means there will be previous year applications still awaiting provider decision. Could we add a rejection step somewhere?

**Review by mnacos — COMMENTED** [review]
The code re-organisation makes it clearer. The only annoyance is previous year applications still awaiting provider decision.

image

File: lib/tasks/generate_test_data.rake (hunk 2 above)
Comment (tijmenb) [reply lib/tasks/generate_test_data.rake:5]: I've added it here: https://github.com/DFE-Digital/apply-for-teacher-training/pull/3190/files#diff-01a8ebf1b37797dc1ed6f0fcb98758a4ba7f9c9b91f1c5d449fa803ede34255cR37

**Comment by mnacos** [comment]
`GenerateTestApplications` fails with `NoMethodError: undefined method `recruitment_cycle_year=' for #<UCASMatch:0x00005602c63142b8>`

**Comment by tijmenb** [comment]
@mnacos that's because you haven't run the migrations, it was added recently.

File: app/workers/generate_test_applications.rb [hunk 3]
```diff
@@ -1,50 +1,83 @@
...
+      RejectApplicationsByDefault.new.call
+    end
   end
 
-  def create(states:, apply_again: false, course_full: false)
-    @test_applications.create_application(
-      courses_to_apply_to: @courses_to_apply_to,
+private
+
+  def create(recruitment_cycle_year: nil, states:, apply_again: false, course_full: false)
```
Comment (mnacos) [inline app/workers/generate_test_applications.rb:44]:
```suggestion
  def create(recruitment_cycle_year:, states:, apply_again: false, course_full: false)
```

Is it worth making `recruitment_cycle_year:` mandatory here, too? We've already made it mandatory for `TestApplications`.

File: app/workers/generate_test_applications.rb [hunk 4]
```diff
@@ -1,50 +1,83 @@
...
+    TestApplications.new.create_application(
       states: states,
+      recruitment_cycle_year: recruitment_cycle_year,
+      courses_to_apply_to: recruitment_cycle_year == 2020 ? courses_to_apply_to_in_2020 : courses_to_apply_to_in_2021,
       apply_again: apply_again,
       course_full: course_full,
     )
   end
+
+  def courses_to_apply_to_in_2020
```
Comment (mnacos) [inline app/workers/generate_test_applications.rb:54]: From what I've seen, 2020 courses are still `exposed_in_find`, so this and the following method could probably be merged into a common `def courses_to_apply_to_in(recruitment_cycle_year)`.

File: app/workers/generate_test_applications.rb [hunk 5]
```diff
@@ -1,50 +1,83 @@
...
-  def create(states:, apply_again: false, course_full: false)
-    @test_applications.create_application(
-      courses_to_apply_to: @courses_to_apply_to,
+private
+
+  def create(recruitment_cycle_year: nil, states:, apply_again: false, course_full: false)
+    TestApplications.new.create_application(
       states: states,
+      recruitment_cycle_year: recruitment_cycle_year,
+      courses_to_apply_to: recruitment_cycle_year == 2020 ? courses_to_apply_to_in_2020 : courses_to_apply_to_in_2021,
```
Comment (mnacos) [inline app/workers/generate_test_applications.rb:48]: Maybe we don't need this hard-coded check, please see comment below.

**Review by mnacos — COMMENTED** [review]
Thanks @tijmenb this error was from the review app (!) You'd expect the review app to have run all the migrations. Running a `db:migrate` on the heroku app fixed it.

PR looks great. Last few minor comments below.

File: app/workers/generate_test_applications.rb (hunk 4 above)
Comment (mnacos) [reply app/workers/generate_test_applications.rb:54]: See also https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/lib/tasks/local_dev.rake#L48

**Comment by tijmenb** [comment]
@mnacos updated PR!

File: app/workers/generate_test_applications.rb [hunk 6]
```diff
@@ -1,50 +1,73 @@
...
+    TestApplications.new.create_application(
       states: states,
+      recruitment_cycle_year: recruitment_cycle_year,
+      courses_to_apply_to: courses_to_apply_to(recruitment_cycle_year),
       apply_again: apply_again,
       course_full: course_full,
     )
   end
+
+  def courses_to_apply_to(year)
```
Comment (mnacos) [inline app/workers/generate_test_applications.rb:54]: sweet method

**Review by mnacos — APPROVED** [review]
:tada:

**Pull request merged**